### PR TITLE
feat(wait-for-http): add wait-for-http

### PR DESCRIPTION
**Pull request by revolunet**
related to https://github.com/SocialGouv/kosko-charts/pull/281/files

### CODE REVIEW

File: wait-for-http/.gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  stage: "Test"
+  extends: .change_rules_wait_for_http
+  image:
+    name: "${CI_REGISTRY_IMAGE}/wait-for-http:${CI_COMMIT_SHA}"
+    entrypoint: ["/bin/sh"]
+  variables:
+    URL: http://invalid
+    RETRIES: 5
+  script: |
+    ! /wait.sh
```
Comment (douglasduteil) [inline wait-for-http/.gitlab-ci.yml:39]:
```suggestion
  script: 
    - ! wait-for-http http://invalid
```

File: wait-for-http/.gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    - Build wait-for-http
+  stage: "Test"
+  extends: .change_rules_wait_for_http
+  image:
+    name: "${CI_REGISTRY_IMAGE}/wait-for-http:${CI_COMMIT_SHA}"
+    entrypoint: ["/bin/sh"]
+  variables:
+    URL: http://free.fr
+  script:
+    - /wait.sh
```
Comment (douglasduteil) [inline wait-for-http/.gitlab-ci.yml:53]:
```suggestion
    - wait-for-http http://free.fr
```

File: wait-for-http/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+FROM curlimages/curl:7.73.0
+
+COPY wait.sh .
```
Comment (douglasduteil) [inline wait-for-http/Dockerfile:3]:
```suggestion
COPY ./bin /bin
```

File: wait-for-http/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+ENV URL $URL
+ENV RETRIES $RETRIES
+
+ENTRYPOINT ["./wait.sh"]
```
Comment (douglasduteil) [inline wait-for-http/Dockerfile:8]:
```suggestion
ENTRYPOINT ["/bin/wait-for-http"]
```

**Review by douglasduteil — CHANGES_REQUESTED** [review]
You should consider `wait.sh` as a real bin in the linux system `/bin/wait-for-http`

File: wait-for-http/Dockerfile (hunk 4 above)
Comment (douglasduteil) [reply wait-for-http/Dockerfile:8]: The env can be given at run time

File: wait-for-http/wait.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+retry=${RETRIES:-120};
+while ! curl -sSfL $URL && [[ $(( retry-- )) -gt 0 ]];
+do
+  echo "Waiting for $URL... ($(( retry )))" ; sleep 5s ;
```
Comment (douglasduteil) [inline wait-for-http/wait.sh:6]:
```suggestion
url=$1
while ! curl -sSfL $url && [[ $(( retry-- )) -gt 0 ]];
do
  echo "Waiting for $url ... ($(( retry )))" ; sleep 5s ;
```

Comment (douglasduteil) [reply wait-for-http/wait.sh:6]: The url can be the first args

**Comment by douglasduteil** [comment]
You might need to move rename the `wait.sh` to `bin/wait-for-http` and `chmod +x` it

**Comment by revolunet** [comment]
most should be fixed

File: wait-for-http/.gitlab-ci.yml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    entrypoint: ["/bin/sh"]
+  variables:
+    RETRIES: 5
+  # expect wait to fail and a retry duration >= 25secs (5*5)
+  script: |
+    start=`date +%s`
+    ! /bin/wait-for-http http://invalid
+    end=`date +%s`
+    runtime=$((end-start))
+    [[ $runtime -gt 25 ]]
```
Comment (douglasduteil) [inline wait-for-http/.gitlab-ci.yml:43]: 25s is a long time waiting for failure nan ?

```suggestion
  variables:
    RETRIES: 1
  # expect wait to fail and a retry duration >= 5secs
  script: |
    start=`date +%s`
    ! /bin/wait-for-http http://invalid
    end=`date +%s`
    runtime=$((end-start))
    [[ $runtime -gt 5 ]]
```

1s is enough to prove that it works no ?

File: wait-for-http/.gitlab-ci.yml [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    - Build wait-for-http
+  stage: "Test"
+  extends: .change_rules_wait_for_http
+  image:
+    name: "${CI_REGISTRY_IMAGE}/wait-for-http:${CI_COMMIT_SHA}"
+    entrypoint: ["/bin/sh"]
+  variables:
+    URL: http://free.fr
+  script:
+    - /bin/wait-for-http
```
Comment (douglasduteil) [inline wait-for-http/.gitlab-ci.yml:71]:
```suggestion
  script:
    - /bin/wait-for-http http://free.fr
```

File: wait-for-http/README.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# wait-for-http
+
+> Wait for some URL to return 200
+
+## With Docker image
+
+```sh
+$ docker run --rm -e URL=http://my.app registry.gitlab.factory.social.gouv.fr/socialgouv/docker/wait-for-http:<version>
```
Comment (douglasduteil) [inline wait-for-http/README.md:8]:
```suggestion
$ docker run --rm registry.gitlab.factory.social.gouv.fr/socialgouv/docker/wait-for-http:<version> http://my.app
```

File: wait-for-http/bin/wait-for-http [hunk 9]
```diff
@@ -0,0 +1,19 @@
+#!/bin/sh
+
+retry=${RETRIES:-120};
+
+# can use environment variable if 1st arg is not provided
+TARGET=${1:-$URL}
```
Comment (douglasduteil) [inline wait-for-http/bin/wait-for-http:6]:
```suggestion
retry=${WAIT_FOR_RETRIES:-120};

# can use environment variable if 1st arg is not provided
TARGET=${1:-$WAIT_FOR_URL}
```

What do you think about prefixing the env ?

File: wait-for-http/bin/wait-for-http [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+if [[ ! "$TARGET" ]]; then
+  echo "Invalid URL : $TARGET"
+  echo "usage: wait-for-http http://app.url"
+  exit 1
+fi
+
+while ! curl -sSfL $TARGET && [[ $(( retry-- )) -gt 0 ]];
+do
+  echo "Waiting for $TARGET... ($(( retry )))" ; sleep 5s ;
```
Comment (douglasduteil) [inline wait-for-http/bin/wait-for-http:16]:
```suggestion
  echo "Waiting for $TARGET... ($(( retry )))" ; sleep ${WAIT_FOR_INTERVAL} ;
```

What do you think about a `WAIT_FOR_INTERVAL` for the interval ?

**Review by douglasduteil — COMMENTED** [review]
Inspired by https://github.com/SocialGouv/docker/blob/master/wait-for-postgres/bin/wait-for-postgres

File: wait-for-http/.gitlab-ci.yml (hunk 6 above)
Comment (revolunet) [reply wait-for-http/.gitlab-ci.yml:43]: there is 5 retries of 5s each so the test prove it retried long enough ? 2 at least should be enough btw

File: wait-for-http/.gitlab-ci.yml (hunk 7 above)
Comment (revolunet) [reply wait-for-http/.gitlab-ci.yml:71]: the first test ensure it fallback to ENV var when the 1st arg is not given

File: wait-for-http/.gitlab-ci.yml (hunk 6 above)
Comment (douglasduteil) [reply wait-for-http/.gitlab-ci.yml:43]: Let's change the interval to make it possible :)

**Review by douglasduteil — APPROVED** [review]
SOLID 🚀

File: wait-for-http/.gitlab-ci.yml (hunk 6 above)
Comment (revolunet) [reply wait-for-http/.gitlab-ci.yml:43]: its now `WAIT_FOR_INTERVAL`

File: wait-for-http/.gitlab-ci.yml [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  needs:
+    - Lint wait-for-http
+    - Build wait-for-http
+  stage: "Test"
+  extends: .change_rules_wait_for_http
+  image:
+    name: "${CI_REGISTRY_IMAGE}/wait-for-http:${CI_COMMIT_SHA}"
+    entrypoint: ["/bin/sh"]
+  variables:
+    RETRIES: 5
```
Comment (douglasduteil) [inline wait-for-http/.gitlab-ci.yml:36]:
```suggestion
    WAIT_FOR_RETRIES: 1
```

File: wait-for-http/.gitlab-ci.yml (hunk 6 above)
Comment (douglasduteil) [reply wait-for-http/.gitlab-ci.yml:43]:
```suggestion
  variables:
    WAIT_FOR_RETRIES: 5
  # expect wait to fail and a retry duration >= 25secs (5*5)
  script: |
    start=`date +%s`
    ! /bin/wait-for-http http://invalid
    end=`date +%s`
    runtime=$((end-start))
    [[ $runtime -gt 5 ]]
```

**Review by douglasduteil — CHANGES_REQUESTED** [review]

File: wait-for-http/.gitlab-ci.yml (hunk 6 above)
Comment (revolunet) [reply wait-for-http/.gitlab-ci.yml:43]: WAIT_FOR_RETRIES * WAIT_FOR_INTERVAL = 25

Comment (douglasduteil) [reply wait-for-http/.gitlab-ci.yml:43]: Nieh... 
I mess it up with the gh suggestion let's do it again

**Review by douglasduteil — APPROVED** [review]

**Comment by douglasduteil** [comment]
@revolunet I commited some changes on the branch directly 🙏

**Pull request merged**